interface/src/app/sample_annotation/annotation.js

```diff
@@ -0,0 +1,129 @@
+/**
+ * "adage.sampleAnnotation" module.
+ */
+
+angular.module('adage.sampleAnnotation', [
+  'ui.router',
+  'ui.bootstrap',
+  'adage.utils'
+])
+
+.config(['$stateProvider', function($stateProvider) {
+  $stateProvider.state('sampleAnnotation', {
+    url: '/sample_annotation?node&samples',
+    views: {
+      main: {
+        templateUrl: 'sample_annotation/annotation.tpl.html',
+        controller: 'SampleAnnotationCtrl as ctrl'
+      }
+    },
+    data: {pageTitle: 'Sample Annotations'}
+  });
+}])
+
+.factory('Activity', ['$resource', 'ApiBasePath',
+  function($resource, ApiBasePath) {
+    return $resource(ApiBasePath + 'activity');
+  }
+])
+
+.factory('Sample', ['$resource', 'ApiBasePath',
+  function($resource, ApiBasePath) {
+    return $resource(ApiBasePath + 'sample');
+  }
+])
+
+.controller('SampleAnnotationCtrl', [
+  '$stateParams', '$q', '$log', 'errGen', 'Activity', 'Sample',
+  function($stateParams, $q, $log, errGen, Activity, Sample) {
+    var self = this;
+    self.queryStatus = 'Connecting to the server ...';
+
+    // Ensure that the URL includes sample ID(s).
+    if (!$stateParams.samples) {
+      self.queryStatus = 'Please specify sample ID(s) in the URL.';
+      return;
+    }
+
+    var samplesInUrl = $stateParams.samples;
+    self.uniqueAnnotationTypes = [];
+    // Note: Ideally self.uniqueAnnotationTypes should be a hashtable (such as
+    // an object of Set) instead of Array, but Set is only available since ES6.
+
+    self.hasNode = false;
+    var sampleDict = {};
+
+    // Function that sets valid sample IDs asynchronously:
+    var getSampleIDPromise = function() {
+      var deferred = $q.defer();
+      // If 'node' is found in the URL, get the activity values based on node
+      // and sample(s):
+      if ($stateParams.node) {
+        self.hasNode = true;
+        self.precision = 5;
+        var nodeID = $stateParams.node;
+        Activity.get(
+          {'node': nodeID, 'sample__in': samplesInUrl},  // parameters of GET
+          function success(response) {
+            var validSamples = [];
+            response.objects.forEach(function(element) {
+              validSamples.push(element.sample);
+              sampleDict[element.sample] = {activity: element.value};
+            });
+            deferred.resolve(validSamples.join());
+          },
+          function error(response) {
+            var errMessage = errGen('Failed to get activities', response);
+            $log.error(errMessage);
+            self.queryStatus = errMessage + '. Please try again later.';
+            deferred.reject(errMessage);
+          });
+      } else { // If 'node' is not in the URL, include all input sample IDs.
+        deferred.resolve(samplesInUrl);
+      }
+      return deferred.promise;
+    };
```

Comment:
I am not really convinced that the version you wrote is more readable. I actually still prefer having a separate getSampleIDPromise() function to make the code more modular.